### PR TITLE
[di] When emitting element addresses to destroy fields, use begin_access [deinit].

### PR DESCRIPTION
**Pull request by gottesmm**

### CODE REVIEW

**Comment by gottesmm** [comment]
@swift-ci test

**Comment by gottesmm** [comment]
@swift-ci benchmark

File: lib/SILOptimizer/Mandatory/DefiniteInitialization.cpp [hunk 1]
```diff
@@ -2303,14 +2303,25 @@ SILValue LifetimeChecker::handleConditionalInitAssign() {
...
+        SmallVector<std::pair<SILValue, EndScopeKind>, 4> EndScopeList;
+        EltPtr =
+            TheMemory.emitElementAddressForDestroy(Elt, Loc, B, EndScopeList);
         if (auto *DA = B.emitDestroyAddrAndFold(Loc, EltPtr))
           Destroys.push_back(DA);
-        while (!EndBorrowList.empty()) {
-          SILValue Borrowed, Original;
-          std::tie(Borrowed, Original) = EndBorrowList.pop_back_val();
-          B.createEndBorrow(Loc, Borrowed, Original);
+        for (auto p : EndScopeList) {
```
Comment (gottesmm) [inline lib/SILOptimizer/Mandatory/DefiniteInitialization.cpp]: Looking at this again, I need to think about if I need to pop off from the back instead to get the ordering right.

Comment (gottesmm) [reply lib/SILOptimizer/Mandatory/DefiniteInitialization.cpp]: Yea, I think I am wrong to do it this way. I think the mistake I saw before was that I was not inserting the cleanup for a borrow into the stack at the appropriate time. For the present it shouldn't be an issue, but I am going to fix before merging.

**Comment by gottesmm** [comment]
@swift-ci test

**Comment by gottesmm** [comment]
@swift-ci test

**Comment by gottesmm** [comment]
@swift-ci benchmark

**Comment by gottesmm** [comment]
@swift-ci benchmark

**Comment by gottesmm** [comment]
@swift-ci benchmark

**Comment by gottesmm** [comment]
@swift-ci test

**Comment by gottesmm** [comment]
@swift-ci benchmark

**Comment by gottesmm** [comment]
@swift-ci test linux platform

**Review by rjmccall — APPROVED** [review]
LGTM

**Comment by gottesmm** [comment]
@swift-ci test Linux platform

**Pull request merged**